src/os/bluestore/BlueStore.cc

```diff
@@ -13974,6 +13986,60 @@ int BlueStore::_deferred_replay()
   return r;
 }
 
+bool BlueStore::_eliminate_outdated_deferred(bluestore_deferred_transaction_t* deferred_txn,
+					     interval_set<uint64_t>& bluefs_extents)
+{
+  bool has_some = false;
+  dout(30) << __func__ << " bluefs_extents: " << std::hex << bluefs_extents << std::dec << dendl;
+  auto it = deferred_txn->ops.begin();
+  while (it != deferred_txn->ops.end()) {
+    PExtentVector new_extents;
+    ceph::buffer::list new_data;
+    uint32_t data_offset = 0;
+    dout(30) << __func__ << " input extents: " << it->extents << dendl;
+    for (auto& e: it->extents) {
+      uint64_t offset = e.offset;
+      uint32_t length = e.length;
+      interval_set<uint64_t> region;
+      region.insert(offset, length);
+
+      auto mi = bluefs_extents.lower_bound(offset);
+      if (mi != bluefs_extents.begin()) {
+	--mi;
+	if (mi.get_end() <= offset) {
+	  ++mi;
+	}
+      }
+      while (mi != bluefs_extents.end() && mi.get_start() < offset + length) {
+	interval_set<uint64_t> temp;
+	region.union_insert(mi.get_start(), mi.get_len());
+	region.erase(mi.get_start(), mi.get_len());
+	++mi;
+      }
+      for (auto ki = region.begin(); ki != region.end(); ki++) {
+	ceph::buffer::list temp;
+	temp.substr_of(it->data, data_offset + ki.get_start() - offset, ki.get_len());
```

Comment:
A little confused why we subtract offset here? I guess we want the relative positioning of the region to data or something?

Reply:
Added comments. I hope it is clear now.